From: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
Date: Fri, 8 Oct 2021 20:31:49 +0200
Subject: [PATCH 10/10] sch_htb: Use helpers to read stats in ->dump_stats().

The read of packets/bytes statistics in htb_dump_class_stats() appears
not to be synchronized. htb_dump_class_stats() does not acquire locks
but I'm not sure if the other `bstats' that are read can be modified or
are stable while this callback in invoked.

Add a helper to read the two members while synchronizing against
seqcount_t.

Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
---
 include/net/sch_generic.h |   16 ++++++++++++++++
 net/sched/sch_htb.c       |   18 +++++++++---------
 2 files changed, 25 insertions(+), 9 deletions(-)

--- a/include/net/sch_generic.h
+++ b/include/net/sch_generic.h
@@ -849,6 +849,22 @@ static inline void _bstats_update(struct
 	u64_stats_update_end(&bstats->syncp);
 }
 
+static inline void bstats_read_add(struct gnet_stats_basic_sync *bstats,
+				   __u64 *bytes, __u64 *packets)
+{
+	u64 t_bytes, t_packets;
+	unsigned int start;
+
+	do {
+		start = u64_stats_fetch_begin_irq(&bstats->syncp);
+		t_bytes = u64_stats_read(&bstats->bytes);
+		t_packets = u64_stats_read(&bstats->packets);
+	} while (u64_stats_fetch_retry_irq(&bstats->syncp, start));
+
+	*bytes = t_bytes;
+	*packets = t_packets;
+}
+
 static inline void bstats_update(struct gnet_stats_basic_sync *bstats,
 				 const struct sk_buff *skb)
 {
--- a/net/sched/sch_htb.c
+++ b/net/sched/sch_htb.c
@@ -1324,12 +1324,10 @@ static void htb_offload_aggregate_stats(
 			if (p != cl)
 				continue;
 
-			bytes += u64_stats_read(&c->bstats_bias.bytes);
-			packets += u64_stats_read(&c->bstats_bias.packets);
-			if (c->level == 0) {
-				bytes += u64_stats_read(&c->leaf.q->bstats.bytes);
-				packets += u64_stats_read(&c->leaf.q->bstats.packets);
-			}
+			bstats_read_add(&c->bstats_bias, &bytes, &packets);
+			if (c->level == 0)
+				bstats_read_add(&c->leaf.q->bstats,
+						&bytes, &packets);
 		}
 	}
 	_bstats_update(&cl->bstats, bytes, packets);
@@ -1356,13 +1354,15 @@ htb_dump_class_stats(struct Qdisc *sch,
 
 	if (q->offload) {
 		if (!cl->level) {
+			u64 bytes = 0, packets = 0;
+
 			if (cl->leaf.q)
 				cl->bstats = cl->leaf.q->bstats;
 			else
 				gnet_stats_basic_sync_init(&cl->bstats);
-			_bstats_update(&cl->bstats,
-				       u64_stats_read(&cl->bstats_bias.bytes),
-				       u64_stats_read(&cl->bstats_bias.packets));
+
+			bstats_read_add(&cl->bstats_bias, &bytes, &packets);
+			_bstats_update(&cl->bstats, bytes, packets);
 		} else {
 			htb_offload_aggregate_stats(q, cl);
 		}
